### PR TITLE
openlineage: add `opt-in` option

### PR DESCRIPTION
**Pull request by JDarDagran**
### Problem statement
Currently you can only disable OpenLineage for all DAGs and tasks or selectively disable by operator class. Users report lack of possibility to do so on DAG/task level instead while opting in.

### Proposed solution
Add `opt_in` option as well as `enable/disable_lineage` methods that work both on DAG and task objects.
Internally it passes constant parameter (Airflow's `airflow.models.Param`) that is next checked in OpenLineage listener.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/openlineage/provider.yaml
Comment (mobuchowski) [inline airflow/providers/openlineage/provider.yaml]: `force_per_dag_opt_in`?

File: airflow/providers/openlineage/provider.yaml
Comment (uranusjr) [inline airflow/providers/openlineage/provider.yaml]: This version number does not look right?

File: airflow/providers/openlineage/provider.yaml
Comment (kacpermuda) [reply airflow/providers/openlineage/provider.yaml]: 1. `explicit_lineage_control`
2. `selective_lineage_control`
3. `lineage_on_demand`
 ?

File: airflow/providers/openlineage/provider.yaml
Comment (JDarDagran) [reply airflow/providers/openlineage/provider.yaml]: Should I leave it empty? When does this option need to be filled out for the provider release?

Comment (JDarDagran) [reply airflow/providers/openlineage/provider.yaml]: Oh, it marks from which Airflow version it's compatible with, not Airflow provider, isn't it?

Comment (kacpermuda) [reply airflow/providers/openlineage/provider.yaml]: For `disabled_for_operators` it's set to `1.1.0`, so `1.6.0` lgtm if it's the next release. I think it's provider version, not Airflow (that, we specify in dependencies section)

File: airflow/providers/openlineage/provider.yaml
Comment (JDarDagran) [reply airflow/providers/openlineage/provider.yaml]: Both `explicit_lineage_control` and  `selective_lineage_control` work for me. `force_per_dag_opt_in` limits scope per DAG (which isn't true, you can control per task too). `lineage_on_demand` sounds misleading - on what demand, e.g. is this lineage whenever you choose to emit it?

Comment (mobuchowski) [reply airflow/providers/openlineage/provider.yaml]: I think the name should indicate what setting it to `true` or `false` will do. `selective control` and `explicit control` mean nothing tbh

Comment (tatiana) [reply airflow/providers/openlineage/provider.yaml]: What about `disabled_unless_decorated` or `enable_decorated_only`?

Comment (tatiana) [reply airflow/providers/openlineage/provider.yaml]: More precise:`disabled_unless_with_context` or `enable_with_context_only`?

Comment (JDarDagran) [reply airflow/providers/openlineage/provider.yaml]: What is the `context` then?

Comment (tatiana) [reply airflow/providers/openlineage/provider.yaml]: Python context manager (`with`). This may be more explicit:
`disabled_unless_with_context_manager` or `disabled_unless_with_ol_context_manager`

File: airflow/providers/openlineage/provider.yaml
Comment (mobuchowski) [inline airflow/providers/openlineage/provider.yaml]:
```
If this setting is enabled, OpenLineage integration won't collect and emit metadata, unless you explicitely enable it per `DAG` or `Task` using  `enable_lineage` contextmanager.
```

File: airflow/providers/openlineage/provider.yaml
Comment (tatiana) [reply airflow/providers/openlineage/provider.yaml]: Along the lines of what @kacpermuda proposed, what about... `selective_enable` or `exclusive_enable`? it's already in the `openlineage` configuration section, so we probably do not need to add lineage to its name

Comment (JDarDagran) [reply airflow/providers/openlineage/provider.yaml]: I changed it to `selective_enable`. Thanks @tatiana for the suggestions!

File: airflow/providers/openlineage/provider.yaml
Comment (JDarDagran) [reply airflow/providers/openlineage/provider.yaml]: Changed as suggested.

File: airflow/providers/openlineage/plugins/listener.py
Comment (tatiana) [inline airflow/providers/openlineage/plugins/listener.py]: ATM we have `_is_disabled` at `openlineage/plugins/openlineage.py`, which I understand will enable/disable the actual listener.

Would it make sense either to rename this `_is_lineage_enabled` to `_is_selective_lineage_enabled` or to rename the one in `openlineage/plugins/openlineage.py` to `_is_openlineage_listener_enabled` - so it's straightforward to distinguish between them?

File: airflow/providers/openlineage/utils/selective_enable.py
Comment (tatiana) [inline airflow/providers/openlineage/utils/selective_enable.py:33]: It is probably worth adding a comment somewhere highlighting that these settings only take effect when the OpenLineage listener is enabled and adding some reference to how/where this is handled.

File: airflow/providers/openlineage/utils/selective_enable.py
Comment (tatiana) [inline airflow/providers/openlineage/utils/selective_enable.py:34]: If the `disabled` configuration is set to True, would it make sense to warn the users via a log message that this will not work?

File: airflow/providers/openlineage/utils/selective_enable.py
Comment (tatiana) [inline airflow/providers/openlineage/utils/selective_enable.py:50]: It may be worth to add some `INFO` or `DEBUG` log message to highlight we're skipping to send lineage events for this

File: airflow/providers/openlineage/utils/selective_enable.py
Comment (JDarDagran) [reply airflow/providers/openlineage/utils/selective_enable.py:34]: The `disabled` configuration is designed for simplicity and minimal overhead. It's intended for situations where a provider is installed but not used. Since enabling specific features while globally disabling the provider creates conflicting behavior, I would avoid introducing additional logging in this mode.

File: airflow/providers/openlineage/utils/selective_enable.py
Comment (JDarDagran) [reply airflow/providers/openlineage/utils/selective_enable.py:33]: That's user-oriented comment that should be and is located in documentation. Code comments are meant to be developer-oriented.

File: airflow/providers/openlineage/utils/selective_enable.py
Comment (JDarDagran) [reply airflow/providers/openlineage/utils/selective_enable.py:50]: Added such debug logs in `is_dag/task_lineage_enabled`.

File: airflow/providers/openlineage/plugins/listener.py
Comment (JDarDagran) [reply airflow/providers/openlineage/plugins/listener.py]: Renamed.

File: tests/providers/openlineage/utils/test_selective_enable.py
Comment (kacpermuda) [inline tests/providers/openlineage/utils/test_selective_enable.py:32]: I think we should also add some test when there are different settings for DAG and Task. F.e. when enabled on DAG level but disabled for one of tasks.

File: airflow/providers/openlineage/plugins/listener.py
Comment (kacpermuda) [inline airflow/providers/openlineage/plugins/listener.py]: After you merge the current main that includes #38033, there will be no need to assign this as an attibute, as it can be added to conf.py and cached.

**Review by mobuchowski — APPROVED** [review]
LGTM, just address one @kacpermuda comment before merging.

File: airflow/providers/openlineage/plugins/listener.py
Comment (JDarDagran) [reply airflow/providers/openlineage/plugins/listener.py]: Moved logic to `conf.py` and related.

File: tests/providers/openlineage/utils/test_selective_enable.py
Comment (JDarDagran) [reply tests/providers/openlineage/utils/test_selective_enable.py:32]: Added test case.

File: airflow/providers/openlineage/utils/utils.py
Comment (uranusjr) [inline airflow/providers/openlineage/utils/utils.py:84]: Is this correct? This function seems to return True when lineage is enabled, but this branch returns True if `selective_enable` returns False.

File: airflow/providers/openlineage/utils/utils.py
Comment (uranusjr) [inline airflow/providers/openlineage/utils/utils.py]: Let’s add a guard so this is only called on operator objects. The error message would be confusing if this function is (incorrectly) used on a foreign object in the current implementation.

File: docs/apache-airflow-providers-openlineage/guides/user.rst
Comment (uranusjr) [inline docs/apache-airflow-providers-openlineage/guides/user.rst]: This needs to be updated.

Comment (JDarDagran) [reply docs/apache-airflow-providers-openlineage/guides/user.rst]: Yes, it does! Thanks for pointing this out.
Updated.

File: airflow/providers/openlineage/utils/utils.py
Comment (JDarDagran) [reply airflow/providers/openlineage/utils/utils.py]: Added.

File: airflow/providers/openlineage/conf.py
Comment (kacpermuda) [inline airflow/providers/openlineage/conf.py]: We can use _CONFIG_SECTION as in all other functions 😄

**Review by uranusjr — APPROVED** [review]

File: airflow/providers/openlineage/conf.py
Comment (JDarDagran) [reply airflow/providers/openlineage/conf.py]: True. Changed.

**Pull request merged**
